Rewrote MapVC to VIP

jaha01 opened the pull request:
Rewrote Map to VIP and realized saving tapped point

jaha01 committed:
Rewrote MapVC to VIP

Rewrote Map to VIP and realized saving tapped point

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -66,4 +67,10 @@ final class DBService {
+
+    func uploadMapPoint(latitude: String, longitude: String) {
+        let parent = ref.child(authService.getUserID()).child(mapPoints)
+        let values = ["latitude":latitude, "longitude":longitude]
```
Не пойму, у тебя тут одна точка будет храниться, или несколько?

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        map.move(with: cameraPosition, animation: YMKAnimation(type: .smooth, duration: 1.0))
+    }
+    
+//    private func addPlacemark(_ map: YMKMap) {
```
Этот кусок еще нужен?

jaha01 replied:
да, с его помощью на карте будут отображены все точки которые я ранее сохранял

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+    
+    private func move(_ map: YMKMap, to point: YMKPoint = YMKPoint(latitude: 79.935493, longitude: 40.327392)) {
```
Вот это что-то очень жесткое, в качестве дефолтного значения. Что означают эти магические числа?

YarMag replied:
Жесткое дефолтное значение не ушло. В таком виде не приму. Строго говоря, я вообще не вижу причин, по которым у этого метода должно быть дефолтное значение.

YarMag replied:
Третий раз уже коммит, а исправления все нет)

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        view.addSubview(mapView)
+        setConstraints()
+        title = "Карты"
+        let tap = UITapGestureRecognizer(target: self, action: #selector(toucheScreen))
```
tapHandler тогда уж. А то просто тап - это событие, скорее.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+                tilt: 0
+            ),
+            animation: YMKAnimation(type: YMKAnimationType.smooth, duration: 5),
+            cameraCallback: nil)
```
Строки 32-42 заслуживают отдельной функции, типа setupMap(). Кстати, вот это - YMKPoint(latitude: 55.669757585559445, longitude: 37.76495471766904) - тоже необязательная штука. Я так понимаю, она нужна просто для того, чтоб хоть куда-то камеру поставить.

Вот тебе небольшая задачка на следующий раз - запросить у юзера его геолокацию и использовать ее как стартовую точку для карты. Подумай:
1. Как это можно сделать, какие сервисы/классы тебе для этого потребуются.
2. Какая должна быть архитектура у этого процесса - кто что будет запрашивать и куда передавать.

YarMag replied:
Вот, кстати, если ты перенесешь CLLocationManager в интерактор, ты частично найдешь ответ на эти вопросы.

YarMag commented on autoService/Map/MapInteractor.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    // MARK: Public methods
+    
+    func saveMapPoint(latitude: String, longitude: String) {
```
А зачем ты широту и долготу в строки конвертировал? С ними обычно работают как с числами.

YarMag commented on autoService/Info.plist:
```diff
@@ -2,6 +2,8 @@
 <plist version="1.0">
 <dict>
+	<key>NSLocationUsageDescription</key>
+	<string>This is Demo Map</string>
```
В качестве тренировки предлагаю пофантазировать и решить, какой бы реально текст ты бы сюда добавил, будь это продукт, готовящийся к релизу. По своему опыту скажу, что иногда разработчикам так и говорят - "Напишите текст сами".

YarMag commented on autoService/Authentification/AlertManager.swift:
```diff
@@ -26,4 +26,17 @@ class AlertManager {
         }
     }
+    
+    public static func showTapInfo(title: String, message: String, completion:  (() -> ())? = nil) {
```
По-моему, вместо передачи отдельно title/message можно воспользоваться одним конфигом, как выше.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    // MARK: - Private methods
+    
+    @objc func toucheScreen(touch: UITapGestureRecognizer) {
```
touchScreen. А лучше handleScreenTap.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        let tappedPoint = YMKScreenPoint(x: Float(touchPoint.x), y: Float(touchPoint.y))
+        let worldPoint = mapView.mapWindow.screenToWorld(with: tappedPoint) ?? YMKPoint()
+
+        AlertManager.showTapInfo(title: "Attention", message: "You have tapped point \(worldPoint.latitude):\(worldPoint.longitude)") { [weak self] in
```
Знаешь, что бросается в глаза? То, что половина надписей у тебя на русском, половина на английском. Экран называется карты, но в заголовке алерта - Attention. Это выглядит странно. 

Нужно будет (не здесь, отдельным PR-ом) протащить в проект локализацию. Там вроде яблоки не так давно какой-то относительно новый способ работы с ней завезли. В общем, надо будет тебе поискать свежие туториалы по запросу "Swift Xcode Localization". Суть локализации в чем - ты пишешь не хардкодную строку с текстом на каком-то языке, а специальный ключ, по которому уже из специальных файликов берется русская или английская строка. В результате при переключении языка у тебя меняются строки сразу во всем приложении.

jaha01 committed:
.

jaha01 committed:
Current user location

What was done:
1) Getting current location in a first map load
2) Go to saved point by click
3) Hide keys in a Info.plist

YarMag commented on autoService/AppDelegate.swift:
```diff
@@ -15,6 +16,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         FirebaseApp.configure()
+        var apiKey: String {
```
Чот брутально ты сделал, computed property прям внутрь метода запихнул) Вынеси тогда уже в статическое свойство какого-нибудь нового класса, типа ApiKeys (у тебя ж несколько апи ключей, кстати). И доставай как-нибудь типа ApiKeys.YandexMaps, ApiKeys.DaData, и т.д.

YarMag commented on autoService/Map/MapStructs.swift:
```diff
@@ -0,0 +1,30 @@
+
+import Foundation
+
+struct AlertMapPoints {
```
Так нельзя. 
Во-первых, косяк в именовании. Структура называется - "Точки карты для алерта". Множественное число. В самой структуре - информация об ОДНОЙ точке. Противоречие. Если уж очень хочется, назови AlertMapPoint. Если структура у тебя хранит информацию об одной сущности, то и название ее должно быть тоже в единственном числе. Вообще, в принципе, во множественном числе типа данных практически никогда не называют. Переменные - да. Массив точек - mapPoints. Это ок. let mapPoints: [MapPoint]. А что такое будет let something: [MapPoints]? Массив массивов? Матрица? Так это и должно выглядеть тогда как [[MapPoint]]. В общем, делать так НЕЛЬЗЯ.
Во-вторых - не надо жадничать, дай каждой структуре собственный файл)
В-третьих - я бы рассмотрел создание типа отдельно для точки как совокупности широты и долготы. Тогда ты этот тип смог бы использовать и в AlertMapPoint, и просто в MapPoint.

YarMag replied:
По третьему пункту - а есть же вроде тип у CoreLocation для этой цели? Не помню только, как называется, там что-то типа CLLocationCoordinate, что ли.

YarMag commented on autoService/Authentification/AlertManager.swift:
```diff
@@ -26,4 +26,53 @@ class AlertManager {
+        topController.present(alert, animated: true)
+    }
+    
+    public static func showTapInfo(config: AlertMapPoints, completion:  @escaping([String]) -> Void) {
```
Название некорректно. Вообще слова "показать информацию о тапе" вообще не настраивают на то, что тут есть что-то, специфичное для карты. showMapTapInfo - еще ладно. Но ты ж смотри, что метод-то у тебя вообще не для показа информации. Он больше для ввода так-то, тут прям инпуты фигурируют.

YarMag commented on autoService/Authentification/AlertManager.swift:
```diff
@@ -26,4 +26,53 @@ class AlertManager {
+            }
+            
+            alert.addAction(UIAlertAction(title: "Сохранить", style: .default, handler: { action in
+                completion([String((name?.text)!), String((description?.text)!)])
```
Тут нужно чуть по-другому сделать. В чем проблема. Ты возвращаешь какой-то абстрактный массив строк. Мне, как программисту, который потенциально будет работать с кодом, вообще непонятно, что тут происходит, пока я не посмотрю исходный код метода. Смотришь на сигнатуру - там комплишн, возвращающий массив строк. Почему массив? А может ли мне прийти пустой массив? А в каком порядке лежат элементы в этом массив и что они обозначают? А что будет, если я случайно перепутаю и возьму не в том порядке? Если бы ты возвращал тут не массив, а структуру с двумя полями, это бы исключило вообще все вопросы. Пустого ничего прийти не может, всегда есть структура. В этом поле лежит имя, в этом поле лежит описание. Все, просто и понятно и не надо угадывать и разбираться в деталях метода.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        setConstraints()
+        title = "Карты"
+
+        if CLLocationManager.locationServicesEnabled() {
```
Идею понял, молодец. Только CLLocationManager - это все-таки менеджер. То есть бизнес-логика. Во всяком случае, получение геоданных юзера - это точно не UI. Поэтому давай для начала перенесем это в интерактор.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,186 @@
+            locationManager.requestWhenInUseAuthorization()
+            locationManager.startUpdatingLocation()
+        }
+        PointsListViewController().delegate = self
```
А вот тут я щас буду долго и страшно ругаться) Как думаешь, почему?)

YarMag replied:
Главное, в билдере правильно все сделал, а тут бяку оставил)

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        interactor.showPointsList()
+    }
+    
+    private func userCurrentLocation() { // not working
```
Что и почему тут не работает - я так не скажу, с яндексовыми картами не работал. Надо будет отдельно сравнить с их документацией.

YarMag replied:
А название для функции, кстати, не очень выбрал. Напомню, что все методы у нас суть действие. Следовательно, их названия всегда представляет собой глаголы. А тут у тебя это не так.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        return button
+    }()
+    
+    let locationManager = CLLocationManager()
```
Почему публичный?

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,186 @@
+        return view
+    }()
+    
+    private let savedPoints: UIButton = {
```
Почему у тебя кнопка называется "сохраненные точки"?

YarMag commented on autoService/Map/PointsList/PointsListInteractor.swift:
```diff
@@ -0,0 +1,31 @@
+    // MARK: Public properties
+    
+    var presenter: PointsListPresenter!
+    var dbService: DBService!
```
Почему var и почему форс анврап?

YarMag commented on autoService/Map/PointsList/PointsListViewController.swift:
```diff
@@ -0,0 +1,85 @@
+        return table
+    }()
+    
+    var points = [MapPoints]()
```
Почему публичное поле?

YarMag commented on autoService/Map/PointsList/PointsListViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "Ponits List"
```
Опечатка

YarMag commented on autoService/Map/PointsList/PointsListViewController.swift:
```diff
@@ -0,0 +1,85 @@
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        delegate?.show(latitude: points[indexPath.row].latitude,
+                       longitude: points[indexPath.row].longitude)
+        dismiss(animated: true, completion: nil)
```
Фактически, контроллер у тебя сейчас принимает решение о своем закрытии самостоятельно, что не совсем корректно. Любое решение о навигации должен принимать интерактор. Следовательно, цепочка вызовов какая-то такая:
1. В контроллере: interactor.didTapPoint()
2. В интеракторе: router.close()
3. В роутере - viewController?.dismiss...

YarMag replied:
Кстати, в более крупных проектах с более сложной архитектурой контроллер даже про делегата знать ничего не должен. Т.е. он просто докладывает интерактору - была выбрана такая-то точка в таблице. А уж что это за точка и что с ней делать - докладывать о ней делегату или нет, просить роутер закрыть контроллер или нет - это все полностью ответственность интерактора. И там бы протокол был не PointListViewControllerDelegate, а PointListModuleDelegate. Т.е, по сути, делегат идет не от имени всего контроллера, а от имени всего модуля (внутри модуля выбрали точку - модуль завершил свою работу и сообщил результат).

У нас отдельной концепции модуля как сущности в проекте нет, поэтому делегата править не надо. Расскажу тебе потом подробнее про всю эту историю, просто чтобы знал.

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -66,4 +68,31 @@ final class DBService {
+            guard let self = self else { return }
+            if let dictionary = snapshot.value as? [String: Any] {
+                self.points = []
+                for (_, value) in dictionary {
```
Тут же просто map можно сделать?

YarMag commented on autoService/Services/NetworkConfig.swift:
```diff
@@ -11,7 +11,18 @@ import Foundation
     
     let url = "https://suggestions.dadata.ru"
-    let apiKey = "14f74f1003dc89e19695bc03c84741f75e11cc50"
+    var apiKey: String {
```
Вот тут computedProperty уже более уместно. Но давай все-таки вынесем в отдельный класс ApiKeys, как писал в одном из предыдущих комментов.

YarMag replied:
Кстати, это тот редкий случай, когда название класса во множественном числе еще хоть как-то прокатывает) Потому что он будет статический, у него только статические поля (т.е. нельзя будет сделать вызов типа let someObj = ApiKeys(), это просто не имеет смысла). И интерфейс доступа типа ApiKeys.yandexMapKey, ApiKeys.dadataKey не выглядит чем-то странным. У нас коллекция ключей, и мы у нее получаем конкретный ключ.

YarMag commented on autoService/Map/MapInteractor.swift:
```diff
@@ -0,0 +1,36 @@
+    }
+    
+    func showPointsList() {
+        router.showPointsList()
```
Вот этот момент я бы еще немножко переработал.

В моем понимании, PointsList - это "тупой" товарищ, который умеет только показывать список. Map у нас все-таки - более центральный товарищ с точки зрения логики работы с точками. Этот модуль должен выступать источником данных.

Опять же, смотри, вот ты показал PointsList на экране. Ты закрыл модуль. У тебя количество точек не изменилось. Но ты хочешь снова открыть боттом шит с PointsList. Получается, у тебя по новой пойдет скачивание данных из Firebase. Закроешь и откроешь еще раз - скачивание начнется в третий раз. А тебе это и не нужно вовсе, данные-то всегда одни и те же. Поэтому логику работы с точками я бы перенес в этот интерактор. Мы их тут один раз скачаем (и будем обновлять каждый раз, как триггернется listener), сохраним внутри интерактора и будем без лишних скачиваний отдавать данные в PointsList. Естественно, тебе надо будет в метод роутера добавить новый аргумент (массив точек), а в билдер PointsList передавать уже два аргумента - делегата (это, кстати, ты правильно делаешь, молодец) и массив точек для отображения. То, что аргумент у билдера будет не один - это тоже нормально.

jaha01 committed:
fixes

YarMag commented on autoService/Map/MapInteractor.swift:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+    
+    func getUserLocation() {
```
Если ты называешь метод get, он по определению должен иметь возвращаемое значение. Аналогично, если ты называешь метод set, то внутри должна происходить модификация чего-то. В данном случае метод ничего не возвращает, следовательно, его нельзя называть get. Больше похоже на onViewDidLoad. Тем более, что ты именно оттуда и вызываешь.

YarMag commented on autoService/Map/MapInteractor.swift:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+        
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
```
Такие вещи лучше в экстеншне писать. По той же причине, по которой мы реализацию датасорса и делегата в контроллере также выносим в экстеншн.

YarMag commented on autoService/Map/MapRouter.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    func showPointsList(points: [MapPoint]) {
+        guard let mapController = viewController as? MapViewController else { return }
+        let tableVC = PointsListBuilder().build(points: points, delegate: mapController)
```
Почему table? С точки зрения модуля Map, ему совершенно фиолетово, находится внутри коллекция или таблица. Или вообще SwiftUI какой-нибудь. Допустим, завтра ты перепишешь там на коллекшн вью - ты ж не пойдешь сюда переменную переназывать.

Тут вполне хватит pointsListVC. Называй переменные не по деталям реализации (чем меньше про них кто-то знает, тем лучше), а по назначению.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,143 @@
+        return view
+    }()
+    
+    private let savedPlacesList: UIButton = {
```
Что-то не можем мы эту кнопку победить) Читаем, что получилось. Кнопка называется "Список сохраненных мест". Все-таки кнопка - это не список, это кнопка. Которая что-то делает. Лучше назвать это как showSavedPlacesButton.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,143 @@
+        setConstraints()
+        title = "Карты"
+
+        interactor.getUserLocation()
```
Да, тут совершенно точно метод должен называться onViewDidLoad.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,143 @@
+        interactor.getUserLocation()
+    }
+    
+    func setupMap(latitude: Double, longitude: Double) {
```
Почему метод публичный?

YarMag replied:
Не вижу фикса или комментария.

YarMag commented on autoService/Map/MapViewController.swift:
```diff
@@ -0,0 +1,143 @@
+// MARK: - PointsListViewControllerDelegate implementation
+extension MapViewController: PointsListViewControllerDelegate {
+    func show(latitude: Double, longitude: Double) {
+        setupMap(latitude: latitude, longitude: longitude)
```
setup - это нечто, что вызывается только один раз, как правило. А эта штука явно предполагает вызов более, чем один раз.

А у тебя тут даже баг из-за этого будет, кстати. Ты будешь бесконечно накидывать обработчики тапов на карту. Как минимум, неплохо бы разделить собственно setup и перемещение карты в другую точку (обычно для этого делают метод moveMap(to point: ...). И если ты их разделишь, то ты сможешь вызывать moveMap с некоторыми координатами и из setup, и отсюда.

YarMag commented on autoService/Map/PointsList/PointsListBuilder.swift:
```diff
@@ -0,0 +1,29 @@
+        let interactor = PointsListInteractor(dbService: DI.shared.dbService())
+        let presenter = PointsListPresenter()
+        let router = PointsListRouter()
+        
```
Две пустых строки подряд, не фэншуй.

YarMag commented on autoService/Map/PointsList/PointsListViewController.swift:
```diff
@@ -0,0 +1,84 @@
+        let table = UITableView()
+        table.delegate = self
+        table.dataSource = self
+        table.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
```
Вот "cell" бы вынести в какую-нибудь приватную константу. Сейчас, если ты ее поменяешь здесь, у тебя сломается метод cellForRow. Надо будет править в двух местах, про что очень легко забыть. А если сразу завязаться на константу, то проблем не будет.

YarMag commented on autoService/Map/Structs/PointTitle.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct PointTitle {
```
Может, PointInfo? А то Title обычно подразумевает один большой заголовок, а у тебя тут и имя, и описание.

YarMag commented on autoService/Services/DBService.swift:
```diff
@@ -66,4 +67,24 @@ final class DBService {
+        ref.child(authService.getUserID()).child(mapPoints).observe(.value) { snapshot in
+            if let dictionary = snapshot.value as? [String: Any] {
+                handler(dictionary.map { MapPoint(dictionary: $0.value as! [String : Any])
+                })
```
На одной строке тогда уж все оставь. А то странно смотрится.

YarMag commented on autoService/Map/PointsList/PointsListBuilder.swift:
```diff
@@ -0,0 +1,29 @@
+        interactor.router = router
+        router.controller = controller
+        controller.delegate = delegate
+        controller.showPoints(list: points)
```
Так, я только щас заметил. Вот это вот нехорошо делать. Роутер не занимается бизнес логикой. Как должно быть.

1. Вью контроллер показывается на экране. Говорит интерактору - я загрузился (onViewDidLoad). 
2. Интерактор такой - ага, он загрузился, а у меня есть точки. Говорит презентеру - пни там, чтобы список показали.
3. Презентер - контроллер, вот тебе точки, показывай.

YarMag replied:
В целом у тебя почти все готово, у контроллера соответствующий метод даже написан. Просто точки нужно положить в интерактор через инит. А showPoints дергать через презентер.

jaha01 committed:
some configurations

YarMag commented on autoService/Map/MapInteractor.swift:
```diff
@@ -34,21 +35,25 @@ final class MapInteractor: NSObject, CLLocationManagerDelegate{
         dbService.setupMapPointsListeners { [weak self] points in
             guard let self = self else { return }
-            self.router.showPointsList(points: points)
+            self.mapPoints = points.map {$0}
```
Эм, а просто self.mapPoints = points нельзя?

jaha01 replied:
согласен фигня

YarMag commented on autoService/Map/PointsList/PointsListBuilder.swift:
```diff
@@ -15,14 +15,13 @@ final class PointsListBuilder {
         router.controller = controller
         controller.delegate = delegate
-        controller.showPoints(list: points)
+        interactor.pointAfterLoad(points: points)
```
Нет. Нельзя в билдере дергать никакой бизнес-логики, никаких методов. Он не для этого. Ни у интерактора, ни у презентера, ни у кого-либо еще. Задача билдера - создать объекты и проставить зависимости, ничего сверх этого он не делает. Я тебе объяснил, как это должно быть - переделай, пожалуйста.

YarMag replied:
Я уже молчу о том, что ты метод назвал существительным. Напоминаю - методы мы называем действиями. Переменные и типы данных - существительными (причем обычно в единственном числе).

jaha01 replied:
"Т.е. ты передаешь точки в билдер, а билдер уже может их спокойно скормить интерактору"
А как тогда скормить?

YarMag replied:
В инит интерактору, я ж писал.

YarMag replied:
Скармливаешь в инит, и интерактор ждет своего часа до тех пор, пока вьюха не сообщит ему, что готова грузиться (onViewDidLoad). В других ситуациях может получиться так, что ты аналогичный метод дернешь раньше времени, он быстро отработает, ничего по факту не покажет - и к моменту, когда вьюха реально будет готова показаться на экране, действие уже пройдет, и ты ничего не увидишь. А потом будешь удивляться - ну вот же, метод отработал, почему на экране нет ничего)

YarMag commented on autoService/Map/PointsList/PointsListViewController.swift:
```diff
@@ -20,12 +20,13 @@ final class PointsListViewController: UIViewController {
     weak var delegate: PointsListViewControllerDelegate?
     
     // MARK: - Private properties
+    private let cell = "cell"
```
cellIdentifier тогда уж. А то - ячейка. Ячейка - это UITableViewCell, или UICollectionViewCell, или наследник от них.

jaha01 committed:
and one more)

hope it'll be last one

jaha01 committed:
так?

jaha01 committed:
made setup private
